src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -372,45 +372,67 @@ private static bool TrySetTargetMethod(ref GenerationContext context, string nam
         {
             TypeDesc targetType = context.TargetType;
 
+            bool skipPrivate = false;
             MethodDesc targetMaybe = null;
-            foreach (MethodDesc md in targetType.GetMethods())
+            do
             {
-                // Check the target and current method match static/instance state.
-                if (context.IsTargetStatic != md.Signature.IsStatic)
+                foreach (MethodDesc md in targetType.GetMethods())
                 {
-                    continue;
-                }
+                    // Respect private visibility for inheritance scenarios.
+                    if (skipPrivate
```

Comment:
So are you saying we can drop the `private` check and focus on a `virtual` check?

Reply:
> As long as we have the same semantics as CoreCLR on the things that CoreCLR accepts, things are fine.

Yes. None of these issues were on my radar because I just assumed CoreCLR's `MethodIterator` and NativeAOT's `TypeDesc.GetMethods()` both had the same semantics. I'm just interested in them matching. Whichever is cheapest is fine with me.

Reply:
> So are you saying we can drop the `private` check and focus on a `virtual` check?

I was thinking we could afford to ignore both visibility and virtualness on the NativeAOT side, but the other comment made me realize that we probably can't ignore either because the condition that we ignored could lead to ambiguity in cases that CoreCLR didn't consider ambiguous.

I think we need to check both visibility and virtualness (match whatever CoreCLR does).